### ISSUE TITLE
How to fully build llvm-libc

### ISSUE THREAD

**Issue by icuxika** [llvm/llvm-project#118982]
I want to build libc according to the [Full Host Build](https://libc.llvm.org/full_host_build.html), but there are some configurations that I don't understand well.

```
SYSROOT=/home/icuxika/CommandLineTools/llvm-fullbuild
cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lldb;lld" -DLLVM_ENABLE_RUNTIMES="libc;compiler-rt;libcxx;libcxxabi;libunwind" -DDEFAULT_SYSROOT=$SYSROOT -DCMAKE_INSTALL_PREFIX=$SYSROOT -DCMAKE_BUILD_TYPE=Release -DLLVM_PARALLEL_COMPILE_JOBS=12 -DLLVM_PARALLEL_LINK_JOBS=4 -DLLVM_LIBC_FULL_BUILD=ON -DLLVM_LIBC_INCLUDE_SCUDO=ON -DCOMPILER_RT_BUILD_SCUDO_STANDALONE_WITH_LLVM_LIBC=ON -DCOMPILER_RT_BUILD_GWP_ASAN=OFF -DCOMPILER_RT_SCUDO_STANDALONE_BUILD_SHARED=OFF
```
> I tried to build using the above command, but it ended up saying that the header file of the standard library could not be found.

##
- llvm version 19.1.5
- os Ubuntu 22.04.5 LTS

## Question
1. Should` SYSROOT` point to an empty directory? Or point to a compiled llvm directory? Or download and unzip the directory from GitHub release?
2. Will not using `lld` and `compiler-rt` affect this build?
3. Will using gcc and g++ affect this build?

## build err log
```
CMakeFiles/clang_rt.builtins-i386.dir/eprintf.c.o
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -O3 -DNDEBUG -m32 -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/eprintf.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/eprintf.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/eprintf.c.o -c /home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/eprintf.c
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/eprintf.c:10:10: fatal error: 'stdio.h' file not found
   10 | #include <stdio.h>
      |          ^~~~~~~~~
1 error generated.
[103/316] Building C object CMakeFiles/clang_rt.builtins-i386.dir/emutls.c.o
FAILED: CMakeFiles/clang_rt.builtins-i386.dir/emutls.c.o
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -O3 -DNDEBUG -m32 -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/emutls.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/emutls.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/emutls.c.o -c /home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/emutls.c
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/emutls.c:10:10: fatal error: 'stdlib.h' file not found
   10 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
[104/316] Building C object CMakeFiles/clang_rt.builtins-i386.dir/enable_execute_stack.c.o
FAILED: CMakeFiles/clang_rt.builtins-i386.dir/enable_execute_stack.c.o
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -O3 -DNDEBUG -m32 -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/enable_execute_stack.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/enable_execute_stack.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/enable_execute_stack.c.o -c /home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/enable_execute_stack.c
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/enable_execute_stack.c:12:10: fatal error: 'sys/mman.h' file not found
   12 | #include <sys/mman.h>
      |          ^~~~~~~~~~~~
1 error generated.
[108/316] Building C object CMakeFiles/clang_rt.builtins-i386.dir/cpu_model/x86.c.o
FAILED: CMakeFiles/clang_rt.builtins-i386.dir/cpu_model/x86.c.o
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -O3 -DNDEBUG -m32 -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/cpu_model/x86.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/cpu_model/x86.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/cpu_model/x86.c.o -c /home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/cpu_model/x86.c
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/cpu_model/x86.c:24:10: fatal error: 'assert.h' file not found
   24 | #include <assert.h>
      |          ^~~~~~~~~~
1 error generated.
[110/316] Building C object CMakeFiles/clang_rt.builtins-i386.dir/clear_cache.c.o
FAILED: CMakeFiles/clang_rt.builtins-i386.dir/clear_cache.c.o
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -O3 -DNDEBUG -m32 -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/clear_cache.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/clear_cache.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/clear_cache.c.o -c /home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/clear_cache.c
/home/icuxika/Source/llvm-project-llvmorg-19.1.5/compiler-rt/lib/builtins/clear_cache.c:11:10: fatal error: 'assert.h' file not found
   11 | #include <assert.h>
      |          ^~~~~~~~~~
1 error generated.
[118/316] Building C object CMakeFiles/clang_rt.builtins-i386.dir/gcc_personality_v0.c.o
ninja: build stopped: subcommand failed.
FAILED: runtimes/builtins-stamps/builtins-build /home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/runtimes/builtins-stamps/builtins-build
cd /home/icuxika/Source/llvm-project-llvmorg-19.1.5/build/runtimes/builtins-bins && /usr/local/bin/cmake --build .
ninja: build stopped: subcommand failed.
```

**Comment by michaelrj-google** [comment]
The quick answers:
1) With the cmake command you have there, removing the `-DDEFAULT_SYSROOT=$SYSROOT` will allow you to build LLVM-libc properly. With the original command, `$SYSROOT` should point to an empty directory you want to install LLVM-libc in. For the command you've written, it would need the LLVM-libc headers to already be installed in `$SYSROOT`.
2) Skipping lld and compiler-rt should still work, but we generally recommend using an LLVM toolchain since that's what we test most thoroughly.
3) The build you're doing is called a runtimes bootstrap build. It will first build clang with your specified compiler, then build the runtimes with that clang. The compiler you pass to cmake doesn't really matter.

If you're just looking to build LLVM-libc with the GNU tools and you don't need the rest of the LLVM toolchain (e.g. clang, compiler-rt, etc.) you should try the build configuration we use for development (under [Configure the build for development](https://libc.llvm.org/full_host_build.html#id3)). You can change the compiler arguments to `gcc` and `g++`, and add `-DCMAKE_INSTALL_PREFIX=/path/to/install/dir` to install just the libc to whatever directory you want. Generally we recommend building with clang but we do make sure that gcc works as well.

The complex answer:
If you specify both `LLVM_ENABLE_PROJECTS` and `LLVM_ENABLE_RUNTIMES` and also root your cmake in the `llvm` directory, you get a bootstrapping build. First it will build `clang` with the compiler you provide to cmake, then it will build your runtimes. However, within those runtimes it first builds the builtins that the compiler assumes exist, which in LLVM come from compiler-rt. Those builtins require the libc headers, and the empty sysroot doesn't have them, so you get the error you saw and the build fails.
This is mostly an ordering problem, but one that's hard to untangle. LLVM-libc depends on the compiler builtins that compiler-rt provides, and compiler-rt depends on the headers that LLVM-libc provides. To break this dependency loop you have two main options: do a separate build just to install the LLVM-libc headers in `$SYSROOT`, at which point your original command should work, or remove the `-DDEFAULT_SYSROOT=$SYSROOT` which will mean your compiler-rt gets built against the system libc headers.
The reason the original command doesn't have this problem is that it puts everything in the `PROJECTS` list, so it can all be built independently. Putting `libc` and `compiler-rt` in the `PROJECTS` list isn't recommended (and in fact will be going away soonish), but in this case it allows you to build the libc headers, then compiler-rt. That being said I don't know why clang itself doesn't have any problem building with an empty sysroot.

Regardless, I appreciate you pointing out a place where our docs are lacking, I'll see if I can get them updated before EoY. I'm also curious about what you're planning to use LLVM-libc for. I'm always happy to hear about people projects using our code.

**Comment by icuxika** [comment]
@michaelrj-google Thank you for your reply. I'm still a novice in C/C++. When I was learning to write some C++ projects before, I liked to use clang and related tools. Recently, I needed to run the program on Linux. I have some knowledge about glibc and musl, and I also happened to know about the llvm-libc project. So I wanted to try using it to build a C project.

I've currently tried to build llvm-libc by putting `libc` and `compiler-rt` in the `PROJECTS` and I haven't encountered the previous problems anymore. Next, I'll try to use it to build projects. Some problems should be due to my insufficient understanding of the details of building projects on Linux. I'll try to solve them. Anyway, thank you. 
```
cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS="clang;lld;libc;compiler-rt" -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLLVM_LIBC_FULL_BUILD=ON -DLLVM_LIBC_INCLUDE_SCUDO=ON -DCOMPILER_RT_BUILD_SCUDO_STANDALONE_WITH_LLVM_LIBC=ON -DCOMPILER_RT_BUILD_GWP_ASAN=OFF -DCOMPILER_RT_SCUDO_STANDALONE_BUILD_SHARED=OFF -DCLANG_DEFAULT_LINKER=lld -DCLANG_DEFAULT_RTLIB=compiler-rt -DDEFAULT_SYSROOT=$SYSROOT -DCMAKE_INSTALL_PREFIX=$SYSROOT -DLLVM_PARALLEL_COMPILE_JOBS=12 -DLLVM_PARALLEL_LINK_JOBS=4
```

**Comment by icuxika** [comment]
I encountered some minor problems when using it. I didn't solve them until I read the documentation again. According to the habit of using CMake, after the construction was completed with `cmake --build build`, I carried out the installation operation in the way of `cmake --install build`. Eventually, I found that some crucial libraries, such as `crt1.o`, `crti.o` and `crtn.o`, hadn't been installed. Finally, I executed the `ninja install-clang install-builtins install-compiler-rt install-core-resource-headers install-libc install-lld` and then it could be used normally.
